gcc/d/d-decls.cc

```diff
@@ -101,7 +101,38 @@ VarDeclaration::toSymbol()
 	  return csym;
 	}
 
-      csym = new Symbol();
+      // Use only same Symbol for VarDeclarations with same mangle
+      if (!this->mangleOverride && isDataseg () && !(storage_class & STCextern))
+	{
+	  static hash_map<nofree_string_hash, Symbol *> duplicate_map(251);
+	  const char *vname = mangle(this);
+	  bool existed;
+
+	  Symbol **entry = &duplicate_map.get_or_insert(vname, &existed);
+	  if (!existed)
+	    {
+	      csym = new Symbol();
+	      *entry = csym;
+	    }
+	  else
+	    {
+	      csym = *entry;
+	      if (!csym->Sloc.equals(loc))
+		{
+		  error ("Found different variables with identical mangle: %s"
+		    " at %s and %s", vname, loc.toChars(), csym->Sloc.toChars());
```

Comment:
> The frontend check is not done for functions:

Then maybe frontend should be made consistent?

> And with this pull request and without the loc check, the function example would no longer generate a linker / assembler error. Instead it would simply hook up all references to foo to the first processed function. With the loc check it errors instead.

I wonder whether we could acheive the same by hooking up `lang_identifier` with `binding_level` where we keep track of all active binding levels `DECL_NAME` is being used in.  And throw an error if it shows twice in the same scope.

I would prefer to keep to the long plan of removing `Symbol` and integrating the contents of it into the common gcc frontend language types.